smarc_bt/src/mission_plan.py

```diff
@@ -473,6 +539,200 @@ def path_to_list(self, path_msg):
+        :param tangent_tol: Numerical tolerance at which we decide the intersections are close enough to consider it a tangent
+        :return Sequence[Tuple[float, float]]: A list of length 0, 1, or 2, where each element is a point at which the circle intercepts a line segment.
+
+        Note: We follow: http://mathworld.wolfram.com/Circle-LineIntersection.html
+        """
+
+        (p1x, p1y), (p2x, p2y), (cx, cy) = pt1, pt2, circle_center
+        (x1, y1), (x2, y2) = (p1x - cx, p1y - cy), (p2x - cx, p2y - cy)
+        dx, dy = (x2 - x1), (y2 - y1)
+        dr = (dx ** 2.0 + dy ** 2.0)**.5
+        big_d = x1 * y2 - x2 * y1
+        discriminant = circle_radius ** 2.0 * dr ** 2.0 - big_d ** 2.0
+
+        if discriminant < 0:  # No intersection between circle and line
+            return []
+        else:  # There may be 0, 1, or 2 intersections with the segment
+            intersections = [
+                (cx + (big_d * dy + sign * (-1 if dy < 0 else 1) * dx * discriminant**.5) / dr ** 2.0,
+                 cy + (-big_d * dx + sign * abs(dy) * discriminant**.5) / dr ** 2.0)
+                for sign in ((1, -1) if dy < 0 else (-1, 1))]  # This makes sure the order along the segment is correct
+            if not full_line:  # If only considering the segment, filter out intersections that do not fall within the segment
+                fraction_along_segment = [(xi - p1x) / dx if abs(dx) > abs(dy) else (yi - p1y) / dy for xi, yi in intersections]
+                intersections = [pt for pt, frac in zip(intersections, fraction_along_segment) if 0 <= frac <= 1]
+            if len(intersections) == 2 and abs(discriminant) <= tangent_tol:  # If line is tangent to circle, return just one point (as both intersections have same location)
+                return [intersections[0]]
+            else:
+                return intersections
+
+    def dubins_mission_planner(self, mission, num_points=2, inside_turn=True):
+        '''
+        Reads the waypoints from a MissionControl message and generates a sampled dubins
+        path between them.
+
+        :param mission: the original MissionControl message
+        :param num_points: the amount of waypoints to keep on each segment between waypoints
+        :param inside_turn: cut the waypoints corners
+        :return MissionControl.msg: new MissionControl message equal to the input one except for the waypoints attribute
+        '''
+
+        rospy.loginfo("Computing dubins path")
+
+        turn_radius = self.bb.get(bb_enums.DUBINS_TURNING_RADIUS)
+        int_radius = self.bb.get(bb_enums.DUBINS_INTERSECTION_RADIUS)
+        vehicle = self.bb.get(bb_enums.VEHICLE_STATE)
+        robot_name = self.bb.get(bb_enums.ROBOT_NAME)
+
+        ll_to_utm_serv = self._get_latlon_to_utm_service()
+
+        # Get the auv's current position to use as the intial point of the dubins path
```

Comment:
this needs to happen when the mission is _started_, not _received_. You can move this logic to get_current_wp, there is already a check for "is this the first ever request of a wp?" in there.